compiler/optimizer/optimizer.go

```diff
@@ -209,137 +321,125 @@ func (o *Optimizer) propagateScanOrder(op dag.Op, parent order.SortKey) (order.S
 		}
 		return order.Nil, nil
 	case *dag.LakeMetaScan, *dag.PoolMetaScan:
 		return order.Nil, nil
-	case *dag.Pass:
-		return parent, nil
 	case *kernel.Reader:
-		return s.SortKey, nil
+		return op.SortKey, nil
 	default:
-		return order.Nil, fmt.Errorf("unknown dag.Source type %T", s)
+		return order.Nil, fmt.Errorf("internal error: unknown source type %T", op)
+	}
+}
+
+func (o *Optimizer) sortKey(id ksuid.KSUID) (order.SortKey, error) {
+	pool, err := o.lookupPool(id)
+	if err != nil {
+		return order.Nil, err
 	}
+	return pool.SortKey, nil
 }
 
-// Parallelize takes a sequential operation and tries to
-// parallelize it by splitting as much as possible of the sequence
-// into n parallel branches. The boolean return argument indicates
-// whether the flowgraph could be parallelized.
+// Parallelize tries to parallelize the DAG by splitting each source
+// path as much as possible of the sequence into n parallel branches.
 func (o *Optimizer) Parallelize(n int) error {
-	replicas := n - 1
-	if replicas < 1 {
-		return fmt.Errorf("bad parallelization factor: %d", n)
+	// Compute the number of parallel paths across all input sources to
+	// achieve the desired level of concurrency.  At some point, we should
+	// use a semaphore here and let each possible path use the max concurrency.
+	if len(o.sources) == 0 {
+		return nil
 	}
-	if replicas > 50 {
-		// XXX arbitrary circuit breaker
-		return fmt.Errorf("parallelization factor too big: %d", n)
+	concurrency := n / len(o.sources)
+	if concurrency < 2 {
+		concurrency = 2
 	}
-	seq := o.entry
-	from, ok := seq.Ops[0].(*dag.From)
-	if !ok {
-		return nil
+	if concurrency > 50 {
+		// arbitrary circuit breaker
+		return fmt.Errorf("parallelization factor too big: %d", n)
 	}
-	trunks := poolTrunks(from)
-	if len(trunks) == 1 {
-		quietCuts(trunks[0])
-		if err := o.parallelizeTrunk(seq, trunks[0], replicas); err != nil {
+	for seq := range o.sources {
+		lister, slicer, rest := matchSource(seq.Ops)
+		// We parallelize the scanning to achieve the desired concurrency,
+		// then the step below pulls downstream operators into the parallel
+		// branches when possible, e.g., to parallelize aggregations etc.
+		parallel, err := o.parallelizeScan(rest, concurrency)
+		if err != nil {
 			return err
 		}
+		if parallel == nil {
+			// Leave the source path unmodified.
+			continue
+		}
+		front := []dag.Op{lister}
+		if slicer != nil {
+			front = append(front, slicer)
+		}
+		// Replace the source path with the parallelized gadget.
+		seq.Ops = append(front, parallel...)
 	}
+	o.findParPullups(o.entry)
+	removePassOps(o.entry)
 	return nil
 }
 
-func quietCuts(trunk *dag.Trunk) {
-	if trunk.Seq == nil {
-		return
-	}
-	for _, op := range trunk.Seq.Ops {
-		if cut, ok := op.(*dag.Cut); ok {
-			cut.Quiet = true
-		}
+func (o *Optimizer) lookupPool(id ksuid.KSUID) (*lake.Pool, error) {
+	if o.lake == nil {
+		return nil, errors.New("system error: lake operation cannot be used in non-lake context")
```

Comment:
This is an internal error, right?
```suggestion
		return nil, errors.New("internal error: lake operation cannot be used in non-lake context")
```